[#7] 레포지터리 구현

canlion opened the pull request:
## 목표
- HashMap 대신 DB를 사용하는 레포지터리 구현

## 추가 목표
- SQL DB와 NoSQL DB의 특성을 고려해 무엇을 사용할지 고려해보자.

## 상세

### SQL DB vs NoSQL DB - [참고](https://www.mongodb.com/resources/basics/databases/nosql-explained/nosql-vs-sql)
* SQL
  * 정형 데이터를 테이블 형태로 저장하며 미리 설정한 스키마와 일치하는 데이터만을 저장할 수 있다.
    * 정형 데이터: 데이터의 타입, 순서, 의미 등 미리 정해진 포맷이 존재하는 데이터
    * 스키마: DB에 저장되는 데이터의 포맷, 테이블 간의 관계 등 DB의 구조를 의미
  * 하나의 서버를 운영하며 서버 자체의 성능을 향상시키는 수직적 확장이 주로 이루어진다.
    * 분산 시스템은 노드간의 데이터 일관성과 무결성, 스키마의 일관성 등을 보장하기 어렵다.
* NoSQL
  * 비정형, 반정형 데이터를 데이터 자체의 포맷 그대로 저장할 수 있다.
    * 비정형 데이터: 정의된 데이터 포맷이 없이 형태가 다양한 데이터
    * 반정형 데이터: 미리 정의된 데이터 포맷은 없되 태그, 마커 등을 통해 구성 요소들을 구분하는 데이터 (ex - JSON, XML, ...)
  * Document DB, key-value DB, graph DB 등 여러 타입이 존재
  * 분산 시스템을 통해 성능을 향상시키는 수평적 확장이 주로 이루어진다.
    * 엄격한 데이터 일관성, 무결성의 보장보다는 성능에 초점을 둔다.

서비스를 제공하는 상황이라면 다음과 같이 SQL, NoSQL DB를 함께 사용하는 것이 좋을 것 같습니다.
* 회원 정보는 SQL DB로 관리
  * 회원 정보는 엄격하게 관리되어야 하며 스키마를 따라 테이블 형태로 관리하기 적절한 형태라 생각됩니다.
* 리뷰, 가게 정보, 메뉴 정보는 NoSQL DB로 관리
  * 가게 정보, 메뉴 정보, 리뷰는 텍스트, 사진, 동영상 등 여러 포맷의 데이터를 포함할 수 있습니다.
  * 로그인 후 리뷰를 작성하는 경우보다는 가게 정보를 찾거나 리뷰를 찾는 트래픽이 훨씬 많을 것이며, 사이트의 주된 목적이 리뷰의 제공이므로 장애 대응 측면에서 분산 시스템이 유리할 것으로 생각됩니다.
  * 가게 정보, 메뉴 정보는 보통 영업일 단위로 변경되므로 실시간으로 일관성과 무결성이 훼손될 상황이 적을 것이라 생각됩니다.

### DB 회원 레포지터리 구현

* Mybatis, JPA - [참고](https://www.elancer.co.kr/blog/view?seq=231)
  * Mybatis: 개발자는 SQL문을 XML로 관리하며, Mybatis는 자바 메서드와 SQL문을 맵핑하여 메서드 실행시 대응되는 SQL문을 실행한다.
    * SQL을 직접 작성해야하지만 SQL 전문가라면 성능 최적화에 유리함.
    * SQL을 직접 작성해야하므로 특정 DB에 종속됨.
  * JPA: 자바 객체와 DB 데이터를 맵핑하며 SQL문을 자동으로 생성한다.
    * SQL을 자동으로 생성하므로 DB에 독립적이며 개발과 유지보수에 유리함.
    * SQL을 자동으로 생성하므로 복잡한 쿼리를 만들기 어렵고, 테스크마다 세세하게 최적화는 불가능하므로 튜닝에 노력이 필요함.

회원 정보로 복잡한 정보를 생성하지는 않을 듯 하여 상대적으로 간편하게 구현할 수 있는 JPA를 선택하였습니다.
* 기존 회원(`Member`) 엔티티에 `@Entity`를 추가해 JPA 엔티티로 변경
* `MySqlMemberRepository` 구현
* 테스트 코드는 아직 작성하지 못했고 포스트맨과 mysql workbench를 통해 동작은 확인하였습니다.

canlion committed:
feat: MySql DB 설정 추가

- build.gradle: MySql 관련 의존성 추가
- application.properties: MySql 연결 설정 추가

canlion committed:
feat: Member 클래스를 JPA 엔티티로 설정

- JPA를 통한 DB 작업을 위해 JPA 엔티티로 변경

canlion committed:
feat: MySql 레포지터리 추가

- MySql DB와 연동해 회원 추가, 제거, 조회를 수행하는 레포지터리 추가

canlion committed:
fix: properties 파일 인코딩 타입 수정

- 한글이 깨지지 않도록 UTF-8로 변경함.

canlion committed:
style: 파일이 빈 라인으로 끝나도록 수정

canlion committed:
test: MemberRepository 공통 테스트 코드 추가

- MemberRepository 인터페이스의 구현체가 공통으로 사용할 테스트 코드
  추가
  - HashMap 등을 이용하며 트랜잭션을 사용하지 않는 구현의 경우 초기화
    메서드 clear가 필요함.

canlion commented:
* 레포지터리 테스트 코드를 추가하였습니다.
  * 인프런 강의 `스프링 DB 2편 - 데이터 접근 활용 기술: 섹션3 테스트` 참고
  * `MemberRepository`를 구현하는 개별 클래스마다 테스트 코드를 두는 대신 공통 테스트 코드를 추가했습니다.
  * MySQL 레포지터리 테스트시에 트랜잭션은 롤백되지만 회원 엔티티의 회원식별번호 베이스가 리셋되지 않아 회원 식별 번호를 지정해야할 경우 하드코딩 대신 DB에 존재하는 회원의 식별 아이디를 꺼내 사용했습니다.

f-lab-lyan commented on api/src/main/java/burgermap/entity/Member.java:
```diff
@@ -1,14 +1,19 @@
 package burgermap.entity;
 
+import jakarta.persistence.*;
```
`*` 말고 전체 import를 볼 수 있으면 좋을 것 같습니다. (IntelliJ설정을 변경해주시면 됩니다.)

canlion replied:
모든 소스 코드에서 wildcard import문을 제거하고 개별 import문으로 교체했습니다.

f-lab-lyan commented on api/src/main/java/burgermap/entity/Member.java:
```diff
@@ -1,14 +1,19 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long memberId;
+    @Column(length = 15)
```
관련된 validation이 어딘가에 들어가야합니다.

canlion committed:
refactor: 회원 서비스 구현체의 클래스명을 변경

- 서비스 구현체 클래스명 변경: MemberServiceImpl -> MemberService
  - 서비스 계층은 보통 하나의 구현을 수정하는 형식이므로 과거에
    인터페이스를 제거하였고 그에 맞춰 클래스명을 수정함

canlion committed:
style: 파일 끝에 빈 라인 추가

canlion committed:
style: whildcard import문을 제거함

- 명시적으로 어떤 클래스가 사용되었는지 알 수 있도록 whildcard(*) import
 문을 제거함
